### PR TITLE
Update Kustomization integration KEP

### PR DESCRIPTION
**Pull request by pwittrock**
- add more background and context.
- revisit specifics of UX.
- provide justification for not using plugins.
- add more alternatives considered.

### CODE REVIEW

**Comment by pwittrock** [comment]
cc @kubernetes/sig-cli-maintainers 
cc @kubernetes/sig-cli-pr-reviews

**Comment by Liujingfang1** [comment]
I documented the security concerns in Kustomize [here](https://docs.google.com/document/d/1FYgLVdq-siB_Cef9yuQBmit0PbrE8lsyTBdGI2eA2y8).
To address those concerns, we will take following actions:
- Delete secret generation [Issue](https://github.com/kubernetes-sigs/kustomize/issues/692).
- File paths may not step up [Issue](https://github.com/kubernetes-sigs/kustomize/issues/693).

 @liggitt @monopole @pwittrock @seans3  @soltysh

File: keps/sig-cli/0031-kustomize-integration.md [hunk 1]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+Do one or more of these.
+
+- [ ] A: `kubectl kustomize -f dir/ | kubectl apply -f -`
+- [ ] B: `kubectl apply -f dir/`
+- [ ] C: `kubectl apply -f dir/kustomization.yaml` (similar to B, but slightly different UX)
+
+### Option A: Kustomize Command
+
+Publish the `kustomize build` command as `kubectl kustomize`.  Update documentation demonstrate using Kustomize as
+`kubectl kustomize -f dir/ | kubectl kustomize -f -`.  Consider Option B as a follow up if there is consensus on
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: The second "kustomize" here should be "apply:"
`kubectl kustomize -f dir/ | kubectl apply -f -`.

It is unfortunate that this requires two separate invocations of _kubectl_.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 2]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+the reason for doing so and trade-offs.
+
+#### Why we like this approach
+
+Kustomize does more than the resource builder in kubectl (e.g. the thing that processes files provided with `-f`).
+
+- It is consistent with how tools that produce Resource Config but exist outside kubectl would integrate with
+ `kubectl apply`
+- It is clear to the user that they are getting a new behavior than when they ran `kubectl apply -f dir/`
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: s/clear/clearer/

File: keps/sig-cli/0031-kustomize-integration.md [hunk 3]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+ which was relatively restrictive in terms of what it did.
+- Publishing it as a separate command keeps what the other kubectl commands do simpler
+- We have talked about moving towards this approach for how we teach users to work with imperative generators -
+ e.g. `kubectl create deployment -o yaml --dry-run | kubectl apply -f -` (or something that doesn't exist yet like
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: It looks like the end of this sentence after "like" is missing.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 4]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+  - **Note:** this case still has friction (e.g. it defaults fields - such as creationTimestamp to null - yuck)
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: Use quotation marks or some formatting for the "creationTimestamp" field name.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 5]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+- It makes more sense to first provide a command (Option A) and then integrate into resource builder (Option B)
+  than it does to do the reverse.
 
-When apply, get or delete is run on a directory, check if it contains a kustomization.yaml file. If there is, apply, get or delete the output of kustomize build. Kubectl behaves the same as current for directories without kustomization.yaml.
+#### Why don't like this approach
 
-#### apply
+- It can't address user friction requiring deeper integration - e.g. produce meaningful line numbers in error
+  messages.
+- Most commands would require the same input pipe - e.g. get, delete, etc would all need pipes
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: s/etc/etc./

File: keps/sig-cli/0031-kustomize-integration.md [hunk 6]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+- We haven't trained user with this pattern
+ - We don't currently follow this pattern by default for other commands such as `kubectl create deployment`- which requires
+   additional flags to output content to pipe `--dry-run -o yaml`
+ - We don't do this for file, dir or url targets - e.g. we don't do `curl something | kubectl apply -f -`
+- When demoed, the UX was considered more complicated than directly integrating into the resource builder
+- It’s more typing for the user
+- Kustomize is the way we are addressing issues with declarative workflows so it should be integrated into them.
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: I have wondered when to capitalize kustomize. Here it's at the beginning of the sentence, but elsewhere it's still written as all lowercase even in this position. That's more of a branding concern, but this caught my eye.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 7]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
 The command visible to users is
+
 ```
 kubectl delete -f <dir>
 ```
 
-### Implementation Details/Notes/Constraints
+#### Why we like this approach
+
+- It is capable of friction that requires deeper integration - such as producing errors referencing line
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: Did you really mean "capable of friction," or should that have been something like, "It removes friction," or "It is capable of reducing friction?"

File: keps/sig-cli/0031-kustomize-integration.md [hunk 8]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+  numbers of the original files (rather than the output files).
+- It integrates will all commands - get, delete, etc - would automatically work.
```
Comment (seh) [inline keps/sig-cli/0031-kustomize-integration.md]: s/etc/etc./

**Comment by Liujingfang1** [comment]
This KEP is for the feature issue #633

File: keps/sig-cli/0031-kustomize-integration.md [hunk 9]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
 * [Alternatives](#alternatives)
 
 [Tools for generating]: https://github.com/ekalinin/github-markdown-toc
 
 ## Summary
-[Kustomize](https://github.com/kubernetes-sigs/kustomize) was developed as a distinct tool from kubectl to rapidly experiment on declarative configuration. The code was developed to solve some long standing, complex kubectl [issues](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/0008-kustomize.md#long-standing-issues). The final step is to close the issues by bringing the code into kubectl. This KEP describes how to do this with consistent kubectl UX.
+
+[Kustomize](https://github.com/kubernetes-sigs/kustomize)
+was developed as a subproject of sig-cli by kubectl maintainers with the goal of addressing
+a collection of issues creating friction for declarative workflows in kubectl (e.g. `kubectl apply`).  The
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: Since Brian nicely pointed out those issues, can we link them here for posterity?

File: keps/sig-cli/0031-kustomize-integration.md [hunk 10]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+goal of Kustomize was to eventually bring this functionality back to kubectl in order to complement
+`kubectl apply` and other declarative workflow commands.
+
+- declaratively generating Resource config
+- declaratively transforming Resource config
+- composing collections of Resource config across files and directories
+- layering the above on top of one another
+
+It is independent of, but complementary to, the *server-side apply* initiative that was started later and targeted
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: Link to https://github.com/kubernetes/enhancements/issues/555 for lazy people like myself :smile:

File: keps/sig-cli/0031-kustomize-integration.md (hunk 9 above)
Comment (soltysh) [reply keps/sig-cli/0031-kustomize-integration.md]: OK, I see you've added them below, mind adding a word they are linked below. People will look for them.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 11]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
-configMapGenerator:
-- name: myJavaServerEnvVars
- literals:   
- - JAVA_HOME=/opt/java/jdk
- - JAVA_TOOL_OPTIONS=-agentlib:hprof
+- Exposing all imperative kubectl commands as declarative options
+- Exposing all kubectl generators and schema-aware transformations
+- Providing simpler alternatives to the APIs for declaring Resource Config (e.g. a simple way to create deployments)
+- Providing a templating mechanism (e.g. for generating Resource Config)
+- Publishing with kubectl any other Kustomize sub-commands besides `build`
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: That's the one I had biggest issues with in the original PR. I'd like to narrow the list of those commands to a specific set. I don't want to end up with nasty commands like `kubectl kustomize edit myresource` which will confuse people with `kubectl edit myresource`, etc.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 12]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+- It is more consistent with UX workflow for commands that work directly off a target
+- It has a cleaner and simpler UX than Option A, fewer steps
+- It integrates the solution directly into the commands with issues instead of tacking it on as separate step
+- Users less likely to get confused and accidentally apply a directory with a kustomization.yaml and have
+  it do the wrong thing (it will do the right thing).
+
+#### Why don't like this approach
+
+- Kustomize does more than the resource builder currently does and this could surprise users
+- Commands are doing more than they were before
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: Not backwards compatible, which is a big deal breaker!

File: keps/sig-cli/0031-kustomize-integration.md [hunk 13]
```diff
@@ -32,122 +33,210 @@ superseded-by:
...
+- Most commands would require the same input pipe - e.g. get, delete, etc would all need pipes
+- We haven't trained user with this pattern
+ - We don't currently follow this pattern by default for other commands such as `kubectl create deployment`- which requires
+   additional flags to output content to pipe `--dry-run -o yaml`
+ - We don't do this for file, dir or url targets - e.g. we don't do `curl something | kubectl apply -f -`
+- When demoed, the UX was considered more complicated than directly integrating into the resource builder
+- It’s more typing for the user
+- Kustomize is the way we are addressing issues with declarative workflows so it should be integrated into them.
+
+### Option B / C: Integration into resource builder
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: I'd like to see these 2 separated, there are distinct pros and cons of each that needs to clearly outlined and currently are missing.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 14]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+      previously brought up issues.  There may be other issues not listed here as well.
+      - https://github.com/kubernetes/kubernetes/issues/13241
+      - https://github.com/kubernetes/kubernetes/pull/14993
+      - https://github.com/kubernetes/kubernetes/pull/14918
+- There is a risk that publishing each command as a separately built binary could cause the aggregate download
+  size of the toolset to balloon.  The kubectl binary is *52M* and the kustomize binary is *31M*.  (extrapolate to
+  30+ commands x 30MB).  Before going down this route, we should consider how to we might want to design a solution
+  and the tradeoffs.    
+
+**AI's for kubectl extensibility:**
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: I'm not convinced these are meant to be here. They make sense in the overall kubectl sense, but not tide with kubectl extensions, maybe we should have a separate KEP that will be discussing just that and will outline the decision we'll make within our SIG.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 15]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+annotations:
+  oncallPager: 800-555-1212
+name: alices-app-sec-c7c5tbh526
+type: Opaque
+```
+
+### Implementation Details/Notes/Constraints
+
+In contrast to `kubectl apply`, which was developed directly in kubernetes/kubernetes and had minimal usage or
+real world experience prior, Kustomize was built as a outside of kubernetes/kubernetes in the
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: s/as a outside/as **an** outside/

File: keps/sig-cli/0031-kustomize-integration.md [hunk 16]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+kubernetes-sigs/kustomize repo.  After `kubectl apply` was published, many issues were uncovered in it that should
+have been discovered earlier.  Implementing Kustomize independently allowed more time for gathering feedback and
+identifying issues.
+
+Kustomize library code will be moved from its current repository (location) to the cli-runtime repository used by
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: Wait, isn't that dependent on the actual decision. The KEP oulines 3 possible options, but doesn't pick a winner, but it should so that we can clearly state further actions like this one.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 17]
```diff
@@ -167,30 +256,151 @@ To get the detailed objects in a kustomization
...
+kubectl.  This will be started after cli-runtime has been fully moved out of staging.
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: Drop that last statement, not relevant.

File: keps/sig-cli/0031-kustomize-integration.md [hunk 18]
```diff
@@ -199,35 +409,50 @@ Most implementation will be in cli-runtime
...
+ - execute kustomize build to get a list of resources
+ - write the output to a StreamVisitor
 - [x] When parsing filename parameters in FilenameParam, look for kustomization directories
 - [ ] documentation:
-  - update the examples in kubectl commands
-  - Improve help messages or documentations to list kubectl subcommands that can work with kustomization directories
+ - update the examples in kubectl commands
+ - Improve help messages or documentations to list kubectl subcommands that can work with kustomization directories
 
 ## Alternatives
```
Comment (soltysh) [inline keps/sig-cli/0031-kustomize-integration.md]: This is now mixed up, alternatives should be proposed next to actual proposals, or we move those proposals that were not picked over here. Now, we have 2 places where alternatives are being discussed, which makes this document hard to follow.

File: keps/sig-cli/0031-kustomize-integration.md (hunk 11 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: Can we just keep this as a non-goal and not do it?

File: keps/sig-cli/0031-kustomize-integration.md (hunk 12 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: Is it not?  The error behavior is a bit different, but you would get an error before.

File: keps/sig-cli/0031-kustomize-integration.md (hunk 14 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: +1

**Comment by pwittrock** [comment]
@soltysh I think we have consensus on the semantics around a subcommand which we want regardless.

I think we need to consider the semantics for implicit cli-runtime integration.  I believe we have consensus that this is something we want, but the precise UX and implementation are still up for debate.

I will put forth that we re-scope this PR to the command only, and follow up with a separate PR to add implicit cli-runtime integration for commands taking `-f`.  Thoughts?

File: keps/sig-cli/0031-kustomize-integration.md (hunk 12 above)
Comment (soltysh) [reply keps/sig-cli/0031-kustomize-integration.md]: The one where it picks `kustomize.yaml` based on name is breaking backwards compatibility. Because beforehand nothing as such would be picked or iirc would fail. With kustomize in it will pick it up and proceed with transformations, that's not backwards compatible. With the option where you explicitly point to `kustomize.yaml` that is fine.

File: keps/sig-cli/0031-kustomize-integration.md (hunk 11 above)
Comment (soltysh) [reply keps/sig-cli/0031-kustomize-integration.md]: Do we need to include it, now that we agree it won't be?

**Comment by soltysh** [comment]
> I will put forth that we re-scope this PR to the command only, and follow up with a separate PR to add implicit cli-runtime integration for commands taking `-f`. Thoughts?

Go for it.

File: keps/sig-cli/0031-kustomize-integration.md (hunk 7 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: +1

File: keps/sig-cli/0031-kustomize-integration.md (hunk 3 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: dropping this

File: keps/sig-cli/0031-kustomize-integration.md (hunk 2 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: Dropping this

File: keps/sig-cli/0031-kustomize-integration.md (hunk 12 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: If it used to fail doesn't anymore does this break backwards compatibility?  If it used to create some resources but fail on the kustomize resource and then exit non-0 is that breaking backwards compatibility?  Guarantees do we provide in the cases where a command fails, are the documented?

Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: Addressing this by pulling into another KEP

File: keps/sig-cli/0031-kustomize-integration.md (hunk 16 above)
Comment (pwittrock) [reply keps/sig-cli/0031-kustomize-integration.md]: Updated

**Comment by pwittrock** [comment]
PTAL

File: keps/sig-cli/0034-kustomize-integration.md [hunk 19]
```diff
@@ -0,0 +1,102 @@
+---
+kep-number: 31
```
Comment (Liujingfang1) [inline keps/sig-cli/0034-kustomize-integration.md]: change it to 34

File: keps/sig-cli/0031-kustomize-integration.md [hunk 20]
```diff
@@ -199,35 +338,50 @@ Most implementation will be in cli-runtime
 - [x] vendor `kustomize/pkg` into kubernetes
 - [x] copy `kustomize/k8sdeps` into cli-runtime
 - [x] Implement a Visitor for kustomization directory which
```
Comment (Liujingfang1) [inline keps/sig-cli/0031-kustomize-integration.md]: We don't need to implement a visitor to kustomization directory since the subcommand `kustomize` we're adding doesn't rely on builder or visitor interface.

**Comment by pwittrock** [comment]
Er, I have a separate KEP but same PR.  LMK how you prefer it.  The separate KEP is provisional, so nothing is set in stone with that one.
